lib/ansible/modules/cloud/amazon/aws_kms.py

```diff
@@ -0,0 +1,299 @@
+  sample: { "role": "add", "role grant": "add" }
+have_invalid_entries:
+  description: there are invalid (non-ARN) entries in the KMS entry. These don't count as a change, but will be removed if any changes are being made.
+  type: boolean
+  returned: always
+'''
+
+# these mappings are used to go from simple labels to the actual 'Sid' values returned
+# by get_policy. They seem to be magic values.
+statement_label = {
+    'role': 'Allow use of the key',
+    'role grant': 'Allow attachment of persistent resources',
+    'admin': 'Allow access for Key Administrators'
+}
+
+# import module snippets
+from ansible.module_utils.basic import AnsibleModule
+
+# import a class, we'll use a fully qualified path
+import ansible.module_utils.ec2
+
+import traceback
+import json
+
+try:
+    import botocore
+    HAS_BOTO3 = True
+except ImportError:
+    HAS_BOTO3 = False
+
+def boto_exception(err):
+    '''generic error message handler'''
+    if hasattr(err, 'error_message'):
+        error = err.error_message
+    elif hasattr(err, 'message'):
+        error = str(err.message) + ' ' + str(err) + ' - ' + str(type(err))
+    else:
+        error = '%s: %s' % (Exception, err)
+
+    return error
+
+def get_arn_from_kms_alias(kms, aliasname):
+    ret = kms.list_aliases()
+    key_id = None
+    for a in ret['Aliases']:
+        if a['AliasName'] == aliasname:
+            key_id = a['TargetKeyId']
+            break
+    if not key_id:
+        raise Exception('could not find alias {}'.format(aliasname))
+
+    # now that we have the ID for the key, we need to get the key's ARN. The alias
+    # has an ARN but we need the key itself.
+    ret = kms.list_keys()
+    for k in ret['Keys']:
+        if k['KeyId'] == key_id:
+            return k['KeyArn']
+    raise Exception('could not find key from id: {}'.format(key_id))
+
+def get_arn_from_role_name(iam, rolename):
+    ret = iam.get_role(RoleName=rolename)
+    if ret.get('Role') and ret['Role'].get('Arn'):
+        return ret['Role']['Arn']
+    raise Exception('could not find arn for name {}.'.format(rolename))
+
+def do_grant(kms, keyarn, role_arn, granttypes, mode='grant', dry_run=True, clean_invalid_entries=True):
+    ret = {}
+    keyret = kms.get_key_policy(KeyId=keyarn, PolicyName='default')
+    policy = json.loads(keyret['Policy'])
+
+    changes_needed = {}
+    assert_policy_shape(policy)
+    have_invalid_entries = False
+    for statement in policy['Statement']:
+        for granttype in ['role', 'role grant', 'admin']:
+            # do we want this grant type? Are we on its statement?
+            # and does the role have this grant type?
+
+            if mode == 'grant' and statement['Sid'] == statement_label[granttype]:
+                # we're granting and we recognize this statement ID.
+
+                if granttype in granttypes:
+                    if clean_invalid_entries and len(list(filter(lambda x: not x.startswith('arn:aws:iam::'), statement['Principal']['AWS']))):
```

Comment:
Can you make this into a couple statements for easier readability? Something like:

```
invalid_entries = [x for x in statement['Principal']['AWS'] 
    if not x.startswith('arn:aws:iam::)]
if clean_invalid_entries and len(invalid_entries):
    for entry in invalid_entries:
        statement['Principal']['AWS'].remove(entry)
```

Or there's always:

```
original_size = len(statement['Principal']['AWS'])
statement['Principal']['AWS'] = [x for x in statement['Principal']['AWS'] 
    if x.startswith('arn:aws:iam::)]
have_invalid_entries = (original_size == len(statement['Principal']['AWS']))
```